lib/ecto/association.ex

```diff
@@ -524,6 +529,17 @@ defmodule Ecto.Association.Has do
   end
 
   @doc false
+  def preload(%{related_key: related_key, owner_key: owner_key} = refl, repo, query, fields, prefix, structs, opts) do
+    ids = structs |> Enum.map(&(Map.get(&1, owner_key)))
+    query = assoc_query(refl, query, Enum.uniq(ids))
+    Ecto.Repo.Preloader.fetch_query(structs, refl, query, prefix, {0, related_key}, fields)
+    |> case do
+      list when is_list(list) -> list
+      query -> repo.all(query, opts)
```

Comment:
When do we get a list and when do we get a query?

Reply:
`fetch_query` can return either, in the case where theres an empty list or a function in the preload, this will actually return the data rather than the query. I forgot to add type specs for `fetch_query`, I'll do that too.

Reply:
Perhaps a better name too now that it's public if I can come up with one.